Experimental support for C++ in TAs

jforissier opened the pull request:
This PR proposes experimental support for C++ in TAs.

This has been requested a couple of time (#1708, #2628), and the answer has always been "not supported", essentially for lack of the basic infrastructure needed by important C++ features (global object constructors/destructors, exception handling). Things have changed quite a bit since that time, so here is a set of patches making it possible to run some C++ code in TAs.

I have performed only limited testing (see my `optee_test` PR: https://github.com/OP-TEE/optee_test/pull/405), and there are some limitations as described in the last commit ("ta: experimental C++ support"). But that should be a good start nonetheless.

Feedback welcome!

jbech-linaro commented:
This is cool Jerome! Probably also quite useful extension to OP-TEE. I'm already aware of some projects that are using C++ code on secure side (not OP-TEE based of course).

jforissier commented:
Github shows the commit is a "weird" order, as it happens sometime... probably because I have rebased and changed patch order quite a bit during the development. This link to my branch is logically ordered however: https://github.com/jforissier/optee_os/commits/ta-c%2B%2B.

jenswi-linaro commented:
It seems that Github (sometimes?) orders commits in a PR by date. So these days I'm usually setting new dates on my commits when I think that it might be needed.

jenswi-linaro commented on lib/libutils/isoc/include/stdio.h:
```diff
@@ -12,6 +12,8 @@ typedef struct _FILE FILE;
 
 int printf(const char *fmt, ...)
                     __attribute__ ((__format__ (__printf__, 1, 2)));
+int sprintf(char *str, const char *fmt, ...)
```
Comment for "libutils: add sprintf()"

This function is dangerous and should not be used. I suppose that some C++ thing needs it to be present but hopefully optimized away before use.
How about adding a warning or deprecated attribute that would show if it's not optimized away?

jforissier replied:
It is used by `libstdc++`:
```
$ nm $(aarch64-linux-gnu-g++ -print-file-name=libstdc++.a) | grep sprintf
                 U sprintf
                 U sprintf
```

I'm fine with adding a `deprecated` attribute, or I can even remove the prototype from the header file and move it to the implementation file (`sprintf.c`) with a comment.

jforissier replied:
I have added `__attribute__((deprecated))`.

jenswi-linaro commented on lib/libutils/isoc/fp.c:
```diff
@@ -0,0 +1,15 @@
+#include <stdio.h>
+
+struct _FILE {
+};
```
What does this mean? I suppose the compiler adds a dummy field to make sense of this. How about being explicit and add a `char dummy` field?

jforissier replied:
Fixed

jenswi-linaro commented on lib/libutils/isoc/fputs.c:
```diff
@@ -0,0 +1,17 @@
+	if (stream != stdout && stream != stderr)
+		abort();
+
+	trace_ext_puts(s);
```
`puts()` would be more accurate, `trace_ext_puts()` doesn't add a `\n` at the end.

jforissier replied:
`fputs()` doesn't add a `\n` either...?

jenswi-linaro replied:
Aha, you're correct, my mistake.

jenswi-linaro commented on core/arch/arm/arm.mk:
```diff
@@ -65,8 +65,16 @@ CFG_CORE_RODATA_NOEXEC ?= n
+# EL1 code is always built with strict alignment because some parts access
+# device memory or run when the MMU is disabled. In both cases unaligned
+# accesses are forbidden.
+CFG_SCTLR_ALIGNMENT_CHECK ?= n
```
Comment for "Disable alignment check by default"

Generally I agree, but I think that our CI testing should still use this.

jforissier replied:
You mean build-tested (Shippable) or runtime-tested (IBART)?

jenswi-linaro replied:
Enabling it on IBART would disable C++ tests on IBART. I guess the benefits of `CFG_SCTLR_ALIGNMENT_CHECK=y` is just to smoke out eventual minor inefficiencies so compared to C++ it's rather low prio.

Let's just maintain two targets (one arm32 and one arm64) on shippable with this.

jforissier replied:
OK, fixed

HernanGatta commented:
This looks great!

I haven't yet digested all this properly, but I wanted to mention what we've been doing with the Open Enclave SDK on OP-TEE in this regard.

Since release 0.7.0 of the SDK, we've had C++ support on ARM64 TAs by effectively building the entire C++ runtime into the TA itself (it's the same thing we do for SGX enclaves).

Instead of linking against the C runtime in the TA Dev Kit, we link against our own copies of musl C, [libunwind](https://github.com/libunwind/libunwind), and LLVM's libcxx and libcxxrt. We then do all the initialization of the runtime upon session open (we don't support TLS though). We have our own version of the TA linker script to get all this to work.

Have you had any trouble catching exceptions that travel through mixed frames? That is, C++ code that calls into C code that calls C++ code that throws, while the handler is in the C++ frames below the C ones.

jforissier commented:
> This looks great!

Thanks :grinning: 

> I haven't yet digested all this properly, but I wanted to mention what we've been doing with the [Open Enclave SDK](https://openenclave.io/) on OP-TEE in this regard.
> 
> Since [release 0.7.0](https://github.com/openenclave/openenclave/blob/master/docs/GettingStartedDocs/OP-TEE/Introduction.md) of the SDK, we've had (preview) C++ support on ARM64 TAs by effectively building the entire C++ runtime into the TA itself (it's the same thing we do for SGX enclaves).
> 
> Instead of linking against the C runtime in the TA Dev Kit, we link against our own copies of musl C, [libunwind](https://github.com/libunwind/libunwind), and LLVM's libcxx and libcxxrt (we still build with GCC; SGX enclaves build with clang by default, and can optionally build with GCC). We then do all the initialization of the runtime upon session open (we don't support TLS though). We have our own version of the TA linker script to get all this to work. We got as far as having the [ONNX runtime](https://github.com/microsoft/onnxruntime) do confidential ML inference in an OP-TEE TA. Was pretty cool!

Cool indeed :wink: I have heard about ML inference use cases from other sources, and for sure it brings some "unusual" requirements to the TA environment in terms of run-time support. Your approach is fine, and is very similar to what the guys I talked to have done (they used GCC, not Clang).

In this PR I tried a different way. Instead of rebuilding the compiler and libraries to adapt to the TA environment, I tried to add features in the OS so that the standard Linux cross-compiler can be used. I think it makes sense because (1) we have been using a Linux-targeted toolchain to build TAs since the beginning, and (2) most importantly, what is needed in terms of OS support is quite standard stuff after all: some TLS relocations and a TLS run-time helper, a function to parse the ELF header, and a few additions to the TA linker script. All that is needed for exception handling. And proper processing of the ELF init/fini sections, which was introduced before and reworked here; that's for the global constructors/destructors. The minimal TLS support will come handy if some day we need to support multi-threading -- I've been told that ML inference, being quite compute-intensive, could benefit from that. It's a different topic, but I'm thinking the OCALL support that you have proposed might help implement a pthreads library.

> I for one look forward to leverage the work you've done here, though I don't yet know how that's going to play out; the SDK does not expect help initializing from other entities on the system.
> 
> Questions:
> 
> 1. Have you had any trouble catching exceptions that travel through mixed frames? That is, C++ code that calls into C code that calls C++ code that throws, while the handler is in the C++ frames below the C ones.

You mean like so? https://github.com/OP-TEE/optee_test/pull/405/commits/ba85249b19aeae899bb38f74c20120fa2bde2285
Works fine as far as I can tell :grinning: 

> 2. Any plans to increase the default shared pool size and/or secure memory at least on QEMU? C++ TAs can get pretty huge pretty quick (the binary, plus heap, plus stack - especially when working with templates); we've run into trouble because there wouldn't be enough space, sometimes in the pool, without dynamic shared memory, sometimes in secure world, while loading/running C++ TAs. We've got patches for OP-TEE, QEMU, and the ATF for that if you're interested.

Sure, please do not hesitate to share.

> 3. Have you considered what would happen if TA developers wish to use different C++ STLs?
> 4. Same as 3., but about different C++ runtimes instead.

No, not really. I'm expecting we can add missing run-time support as required. In fact it is the approach I have taken for this PR. I have written a couple of basic C++ test cases and tried to build and run them. Each time I met an issue, I figured out what was missing and implemented it...

> As far as I can tell from this PR, the expectation is that developers would link against GCC's (and later clang's) C++ runtime, STL, and unwinder, right?

Indeed.

> Again, this is really cool stuff!

henrydcase commented:
this is great stuff!
going further - would you consider RFC for support of Rust in TAs? It sounds like natural choice for secure apps

jenswi-linaro commented:
> would you consider RFC for support of Rust in TAs? 

Yes, I think so. At Linaro Connect 2019 in San Diego there was a talk from someone that had just done that. I haven't looked at the code though.

Emantor commented:
Rust on OP-TEE should AFAIR be https://github.com/mesalock-linux/rust-optee-trustzone-sdk

jforissier commented:
> > would you consider RFC for support of Rust in TAs?
> 
> Yes, I think so. At Linaro Connect 2019 in San Diego there was a talk from someone that had just done that. I haven't looked at the code though.

+1

We need to be careful about code size though. I mean, if supporting more languages requires to add a lot of stuff in the runtime code (ldelf, libutee) it could be problematic for some platforms that are tight on memory. But we can certainly deal with that at compile time if needed.

jforissier commented:
Rebased onto master but review commits not squashed yet.

jforissier commented on .shippable.yml:
```diff
@@ -48,7 +48,7 @@ build:
     - _make CFG_SECURE_DATA_PATH=y
     - _make CFG_REE_FS_TA_BUFFERED=y
-    - _make PLATFORM=vexpress-qemu_armv8a CFG_ARM64_core=y CFG_CORE_ASLR=y
+    - _make PLATFORM=vexpress-qemu_armv8a CFG_ARM64_core=y CFG_CORE_ASLR=y CFG_SCTLR_ALIGNMENT_CHECK=y
```
How? Commit "Disable alignment check by default" sets it to `n` by default and I don't see that vexpress sets it?

etienne-lms replied:
`CFG_SCTLR_ALIGNMENT_CHECK ?= y` since aaaf00a257488d2b6fbf8a4e842e1333529c704c and unmodified since.
It is not set from any **build.git** makefiles.

jforissier replied:
The change you are commenting on is in patch "[Review] Disable alignment check by default" which updates "[Disable alignment check by default](https://github.com/OP-TEE/optee_os/pull/3693/commits/11e0b15fb1f8fbcc6b72bbed497aa7bc1d07cd40)" which sets `CFG_SCTLR_ALIGNMENT_CHECK ?= n` globally ;-) As mentioned the libgcc_eh.a library is not compiled with `-mstrict-align`.

etienne-lms replied:
sorry. discard my comment :confused:

jforissier commented:
Squashed/rebased.

The commit order still looks weird although I have used `git rebase --ignore-date`. All commit dates and author dates are now identical (in the same second), and that still confuses GitHub apparently :-/

jenswi-linaro commented:
> The commit order still looks weird although I have used `git rebase --ignore-date`. All commit dates and author dates are now identical (in the same second), and that still confuses GitHub apparently :-/

Yes, that's why I try to separate the dates with at least one second.

jforissier commented:
> > The commit order still looks weird although I have used `git rebase --ignore-date`. All commit dates and author dates are now identical (in the same second), and that still confuses GitHub apparently :-/
> 
> Yes, that's why I try to separate the dates with at least one second.

Right, that's what I did with my latest rebase. Here's the magic incantation I used :wink:
```
git rebase --force-rebase --exec "sleep 2" --exec "git commit --amend --no-edit --date=now" master
```

I suppose we could find a way without sleeping but it's good enough.

jforissier commented on ldelf/include/elf_common.h:
```diff
@@ -656,6 +656,7 @@ typedef struct {
 #define	R_AARCH64_GLOB_DAT	1025	/* Set GOT entry to data address. */
 #define	R_AARCH64_JUMP_SLOT	1026	/* Set GOT entry to code address. */
 #define	R_AARCH64_RELATIVE	1027	/* Add load address of shared object. */
+#define R_AARCH64_TLS_TPREL	1030	/* Offset of the TLS block in the TCB */
```
Agreed (although it is quite stupid to use a tab after `#define`)

jforissier commented on lib/libutils/isoc/sprintf.c:
```diff
@@ -0,0 +1,19 @@
+ */
+
+#include <stdio.h>
+#include <printk.h>
```
yes

jforissier commented on ta/arch/arm/ta.ld.S:
```diff
@@ -44,15 +44,22 @@ SECTIONS {
 	.dynstr : { *(.dynstr) }
 	.hash : { *(.hash) }
 
+
```
I will

jforissier commented on core/arch/arm/kernel/thread_a64.S:
```diff
@@ -858,6 +864,8 @@ END_FUNC el0_sync_abort
 
 	/* Store original sp_el0 */
```
Well, the added code to store tpidr_el0 does not really need any comment IMO, but I agree it looks a bit strange to have a comment for sp_el0 and one for x4..x30 but none for tpidr_el0. I will add one.

jforissier commented on lib/libutee/arch/arm/tcb.c:
```diff
@@ -0,0 +1,206 @@
+ *
+ * Maintains an array of pointers to TLS data for each module in the TCB. Each
+ * module that has a TLS segment has an entry (and consequently, some space in
+ * the tcb_head::tls buffer). The index it the "module ID".
```
Will fix

jforissier commented:
Squashed/rebased with tags applied. `[RFC]` tag removed.

jforissier commented:
Any chance we could get this reviewed and merged into the next release? There has been positive feedback, as well as a more detailed review by @etienne-lms (thanks!) but I'd like some more eyes on this.

@jenswi-linaro what do you think?

I also realized that commit "ldelf, libutee: rework support of DT_INIT_ARRAY/DT_FINI_ARRAY" somehow breaks the TA ABI because `ldelf` no longer exports `__init_fini_info` to the TA. It only concerns TAs built with OP-TEE 3.9 and which have initialization and/or finalization functions. Typically not used by the C compiler, I don't think it is likely to be a problem unless people have been playing with C++ and introducing some kind of support for it, a case which is handled by this PR. But still, perhaps we should maintain compatibility? It is not hard to do, it just adds 4 Kb to `ldelf_hex.o` and therefore to the TEE core. I'll push an update and we can decide to take it or not.

jenswi-linaro commented on ta/arch/arm/ta.ld.S:
```diff
@@ -28,7 +28,11 @@ SECTIONS {
+		*(.eh_frame_hdr)
+		*(.eh_frame_entry .eh_frame_entry.*)
+	}
+	.eh_frame : { KEEP(*(.eh_frame)) *(.eh_frame.*) }
```
Is the `KEEP()` directive mandatory?

jforissier replied:
I think so. It is present in the GCC default linker script (see output of `arm-linux-gnueabihf-gcc -Wl,-verbose`) and it means the symbols in the specified input section have to be kept even when `--gc-sections` is given and the symbols are not referenced.

Now I must admit there is a bit of cargo-cult programming here, because I can't justify why `KEEP()` is not equally applied to `.eh_frame.*` ;-)

jenswi-linaro replied:
> Now I must admit there is a bit of cargo-cult programming here, because I can't justify why KEEP() is not equally applied to .eh_frame.* ;-)

Aha, so that's what's smelling. ;-)

How about dropping this until something breaks and we know why it's needed?

jforissier replied:
This comment I have not addressed. I think we'd rather trust GCC here, our usage is no different as far as I can tell. Applications that produce `.eh_frame` info during compilation need to have them linked into the final binary so that the C++ runtime can find them (using `dl_iterate_phdr()`).

jforissier replied:
Note, I can confirm that if we don't add a `.eh_frame` output section to our linker script, C++ exceptions don't work. I was just commenting on the KEEP() thing.

jenswi-linaro replied:
OK, thanks for looking into this.

jenswi-linaro commented on ldelf/ta_elf_rel.c:
```diff
@@ -120,25 +129,55 @@ static TEE_Result resolve_sym_helper(uint32_t hash, const char *name,
-		return resolve_sym_helper(hash, name, val, elf);
+	if (elf) {
+		/* Search global symbols */
+		if (!resolve_sym_helper(hash, name, val, elf, false, NULL))
```
I woulnd't mind a `false/*!weak_ok*/` instead of just `false`.

jforissier replied:
Sure

jenswi-linaro commented on ldelf/ta_elf_rel.c:
```diff
@@ -120,25 +129,55 @@ static TEE_Result resolve_sym_helper(uint32_t hash, const char *name,
+		if (!resolve_sym_helper(hash, name, val, elf, false, NULL))
+			return TEE_SUCCESS;
+		/* Search weak symbols */
+		if (!resolve_sym_helper(hash, name, val, elf, true,
```
`true/*weak_ok*/` instead of `true`?

jforissier replied:
OK

jenswi-linaro commented on ldelf/ta_elf_rel.c:
```diff
@@ -35,19 +35,26 @@ static uint32_t elf_hash(const char *name)
 			  unsigned int st_type, size_t st_shndx,
 			  size_t st_name, size_t st_value, const char *name,
-			  vaddr_t *val)
+			  vaddr_t *val, bool weak_ok, bool *found_weak_undef)
```
It seems that `found_weak_undef` isn't really needed. Each time this is set to `true` `*val` is assigned `0` by the caller.

jforissier replied:
True, I will simplify.

jenswi-linaro commented on ldelf/ta_elf_rel.c:
```diff
@@ -194,23 +233,47 @@ static void e32_process_dyn_rel(const Elf32_Sym *sym_tab, size_t num_syms,
+	*mod = NULL;
+	e32_get_sym_name(sym_tab, num_syms, str_tab, str_tab_size, rel, &name);
+	ta_elf_resolve_sym2(name, NULL, mod, NULL);
+	if (!*mod)
```
This depends on the supplied `*mod` to be NULL initialized. How about checking the return value first?

jforissier replied:
ok

jenswi-linaro commented on ldelf/ta_elf_rel.c:
```diff
@@ -194,23 +233,47 @@ static void e32_process_dyn_rel(const Elf32_Sym *sym_tab, size_t num_syms,
+	const char *name = NULL;
+
+	e32_get_sym_name(sym_tab, num_syms, str_tab, str_tab_size, rel, &name);
+	ta_elf_resolve_sym2(name, val, &mod, NULL);
```
It seems that `&mod` doesn't need to be supplied for `ta_elf_resolve_sym2()` to work if we're not going to use that value.

jforissier replied:
So true! Now fixed.

jenswi-linaro commented on ldelf/ta_elf_rel.c:
```diff
@@ -429,6 +429,37 @@ static void e64_process_dyn_rela(const Elf64_Sym *sym_tab, size_t num_syms,
+			     const char *str_tab, size_t str_tab_size,
+			     Elf64_Rela *rela, const char **name)
+{
+	size_t sym_idx = 0;
```
This looks like copy&paste from `e64_process_dyn_rela()`. Perhaps we can refactor instead?

jforissier replied:
Agreed

jenswi-linaro commented on ldelf/ta_elf_rel.c:
```diff
@@ -429,6 +429,37 @@ static void e64_process_dyn_rela(const Elf64_Sym *sym_tab, size_t num_syms,
+
+	e64_get_sym_name(sym_tab, num_syms, str_tab, str_tab_size, rela, &name);
+	ta_elf_resolve_sym2(name, &symval, &mod, NULL);
+	*where = symval + mod->tls_tcb_offs + rela->r_addend;
```
This depends on `mod` not being `NULL`. Is that guaranteed?

jforissier replied:
I suppose it could. For instance `extern __thread int foo;` and `foo` is not found in any ELF at runtime.
I'll use `resolve_sym()` everywhere a symbol is expected and a failure should be fatal.

jenswi-linaro replied:
So we'll fail before reaching this point? My point is that it's better with an error message that gives some clue than a crashing ldelf.

jforissier replied:
Yes. Now I'm using `resolve_sym()` to find `mod`, which will `err(res, "Symbol %s not found", name);` if the symbol is not found. In fact I have updated my code to use `resolve_sym()` everywhere because a symbol not found is always a fatal error when processing relocations.

jenswi-linaro commented on ldelf/ta_elf.c:
```diff
@@ -24,6 +25,21 @@
+ * Layout of a 32-bit struct dl_phdr_info for a 64-bit ldelf to access a 32-bit
+ * TA
+ */
+struct _dl_phdr_info32 {
```
Why the leading `_`?

jforissier replied:
No good reason really... This is an internal structure while `dl_phdr_info` is an exported type available to TAs via `<link.h>`.  I felt I had to tag it somehow but it's not useful in any way in fact ;-)

jenswi-linaro commented on ldelf/ta_elf.c:
```diff
@@ -1567,3 +1650,120 @@ TEE_Result ta_elf_set_init_fini_info(bool is_32bit)
+		dlpi->dlpi_phdr = elf->phdr;
+		dlpi->dlpi_phnum = elf->e_phnum;
+		dlpi->dlpi_adds = 1; /* No unloading on dlclose() currently */
+		dlpi->dlpi_subs = 0; /* No unloading on dlclode() currently */
```
s/dlclode/dlclose/

jforissier replied:
yep

jenswi-linaro commented on lib/libutee/arch/arm/user_ta_entry.c:
```diff
@@ -33,42 +35,118 @@ extern struct ta_head ta_head;
+	_num_mod_init += mod;
+
+	TAILQ_FOREACH_SAFE(qe, &_phdr_info_head, link, qe2)
+		free(qe);
```
This works, but looks a bit scary with the dangling pointers. Normally I'd write it like:
``` C
TAILQ_REMOVE(&_phdr_info_head, qe, link);
free(qe);
```
and skipping the cleaning `TAILQ_INIT()` below.

jforissier replied:
Agreed.

jforissier commented on ldelf/ta_elf.c:
```diff
@@ -431,6 +443,8 @@ static void parse_load_segments(struct ta_elf *elf)
 				elf->exidx_start = phdr[n].p_vaddr;
 				elf->exidx_size = phdr[n].p_filesz;
+			} else if (phdr[n].p_type == PT_TLS) {
+				elf->tls_mod_id = next_tls_mod_id(elf);
```
Sure.

jenswi-linaro commented on ldelf/ta_elf.c:
```diff
@@ -1430,9 +1434,268 @@ TEE_Result ta_elf_add_library(const TEE_UUID *uuid)
+	res = ta_elf_resolve_sym("__init_fini_info", &info_va, NULL);
+	if (res) {
+		if (res == TEE_ERROR_ITEM_NOT_FOUND) {
+			/* Older TA */
```
Older or newer TA?

jforissier replied:
Both, actually ;-) The TA is either built before the symbol was introduced (in 3.9.0), or after it was deprecated (by this PR).
I'll fix the comment.

jforissier commented on ldelf/ta_elf_rel.c:
```diff
@@ -143,43 +148,77 @@ static TEE_Result resolve_sym_helper(uint32_t hash, const char *name,
  */
-TEE_Result ta_elf_resolve_sym(const char *name, vaddr_t *val,
-			      struct ta_elf *elf)
+static TEE_Result ta_elf_resolve_sym2(const char *name, vaddr_t *val,
```
No it wouldn't, will fix.

jenswi-linaro commented:
I've gone through the patches, only some minor comments.

jforissier commented on ldelf/ta_elf.c:
```diff
@@ -445,16 +445,25 @@ static void parse_load_segments(struct ta_elf *elf)
 				elf->exidx_size = phdr[n].p_filesz;
 			} else if (phdr[n].p_type == PT_TLS) {
 				elf->tls_mod_id = next_tls_mod_id(elf);
+				elf->tls_start = phdr[n].p_vaddr;
```
In fact `tls_start` etc. should not be needed for a 32-bit binary so I will try to drop this.

jforissier commented on lib/libutee/arch/arm/user_ta_entry.c:
```diff
@@ -33,42 +35,118 @@ extern struct ta_head ta_head;
+	TAILQ_ENTRY(phdr_info) link;
+};
+
+static TAILQ_HEAD(phdr_info_head, phdr_info) _phdr_info_head =
```
Good catch, will fix

jforissier commented:
@jenswi-linaro many thanks for your comments :+1:
I have addressed them all but it was too impractical to just add fixup commits. So I have force pushed a new series. I saved the previous state of this PR in my branch [`ta-c++-20200805`](https://github.com/jforissier/optee_os/tree/ta-c%2B%2B-20200805) for easy comparison.
@etienne-lms I have kept your Acked-by's since I have made only minor updates.

jenswi-linaro commented:
Thanks @jforissier. I think we're ready to go. :-)
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

lxxstone commented:
Hi @jforissier ,

This is great!

But with these TA-C++ patches, I have not yet been able to compile TA by using library functions such as string or vectors in C++.
Is it currently possible to run is what you mentioned, the basic infrastructure needed by important C++ features (global object constructors/destructors, exception handling)?

Thanks

jforissier commented:
Hi @lxxstone,

Constructors and destructors for global objects should work fine now. As for exception handling, it should be OK too provided that everything is linked statically into a single executable TA (i.e., no shared library).

That being said, I think C++ support has not seen much use yet, and the tests that are included in the OP-TEE test suite (optee_test) are quite basic. So I wouldn't be surprised if more work remains to be done. Feel free to open issues here if you find bugs and we will try to address them!

sudopk commented:
This is great. Thank you.

What is the OP-TEE version where this support is included?

Is there any sample TA where I can see this being used as an example?

jforissier commented:
> This is great. Thank you.
> 
> What is the OP-TEE version where this support is included?

3.10.0

> Is there any sample TA where I can see this being used as an example?

Not that I know of. You can find very basic non-regression tests in https://github.com/OP-TEE/optee_test/blob/master/ta/os_test/cxx_tests.cpp.

zz-fz-john commented:
@jforissier  That's cool !,but I wonder if optee  supports class inheritance.

oldoldcoder commented:
Hi, @jforissier I try to use CPP language in TA. For now, thank you for your contribution.

But it seems that I am still more suitable to convert CPP to C language (without using too many features of CPP, because I have not seen a suitable Optee_example that can let me know how to modify the nested makefile file)

jforissier commented:
@oldoldcoder if you can use plain C instead of C++, please do. C++ support is still somewhat limited and requires a specific version of GCC (most recent versions don't work).

oldoldcoder commented:
@jforissier Thank you for your reply. I have found another approach using C that meets my requirements

alaouiz commented:
> not yet been able to compile TA by using library functions such as string or vectors in C++.
> Is it currently possible to run is what you mentioned, the basic infrastructure needed by important C++ features (global object constructors/destructors, exception handling)?

@jforissier, great work. Is it now possible to use std string, maps with C++ ?

anqiAngel commented:
Hi,@oldoldcoder May I ask how you did it? Can you add your contact information if possible? 

> @jforissier Thank you for your reply. I have found another approach using C that meets my requirements

TephrocactusMYC commented:
Hi, experts,

I would like to know the current level of support for C++. Are there any future development plans for C++?

jforissier commented:
@TephrocactusMYC nothing new I'm afraid. The current level of support is that basic C++ code works with GCC 11 (aarch64-linux-gnu-gcc and arm-linux-gnueabihf-gcc) but not later because later versions depend on glibc features that are not implemented in OP-TEE's C library (libutils).

jenswi-linaro commented:
@TephrocactusMYC, patches are welcome.

TephrocactusMYC commented:
> @TephrocactusMYC nothing new I'm afraid. The current level of support is that basic C++ code works with GCC 11 (aarch64-linux-gnu-gcc and arm-linux-gnueabihf-gcc) but not later because later versions depend on glibc features that are not implemented in OP-TEE's C library (libutils).

@jforissier So, for slightly more complex situations, such as class inheritance, function overloading, operator overloading, etc., the support is not very complete, right? Regarding glibc, I wonder if it's possible to replace GCC with musl or other compilers?

And about the runtime issue, is it caused by the open-source license?

jforissier commented:
> @jforissier So, for slightly more complex situations, such as class inheritance, function overloading, operator overloading, etc., the support is not very complete, right?

To the best of my knowledge, all the above work. If I remember correctly, one thing that doesn't work well is exception handling when shared libraries are used.

> Regarding glibc, I wonder if it's possible to replace GCC with musl or other compilers?

"Replacing GCC with musl" doesn't make sense. I suppose you mean "use a GCC toolchain built with (and for) musl libc". It could be an option. I have never tried it. Likely some adaptation would be needed in OP-TEE (libutils/libutee) but with a bit of luck it might work out well. Normally, toolchains are built with a specific target environment in mind (for example: linux with glibc or linux with musl, etc.). Re-using a toolchain that is not specifically made for OP-TEE user space (and instead adapt OP-TEE user space to the toolchain) can be challenging.

> And about the runtime issue, is it caused by the open-source license?

Not at all. As I said: "later [GCC] versions depend on glibc features that are not implemented in OP-TEE's C library (libutils)".

TephrocactusMYC commented:
> To the best of my knowledge, all the above work. If I remember correctly, one thing that doesn't work well is exception handling when shared libraries are used.

This is good news, but it seems like the remaining work will be tricky.

> "Replacing GCC with musl" doesn't make sense. I suppose you mean "use a GCC toolchain built with (and for) musl libc". It could be an option. I have never tried it. Likely some adaptation would be needed in OP-TEE (libutils/libutee) but with a bit of luck it might work out well. Normally, toolchains are built with a specific target environment in mind (for example: linux with glibc or linux with musl, etc.). Re-using a toolchain that is not specifically made for OP-TEE user space (and instead adapt OP-TEE user space to the toolchain) can be challenging.

You are right, that's what I meant. It might be worth a try.
